### PR TITLE
Allow updating yso keywords

### PR DESCRIPTION
**Pull request by Rikuoja**
The point of this PR is to allow YSO vocabulary stay up to date automatically.

Any deprecated YSO keywords are retained in events for backwards compatibility, but mapped to current ones automatically. Any unknown deprecated YSO keywords cause an Exception that will have to be handled manually before automatic YSO updates can continue. This should be no problem, as YSO vocabulary doesn't need to be constantly up to date.

Also, the keyword endpoint is by default set to serve API users by only providing current YSO keywords, in decreasing order of usage, so that the most relevant keywords are always listed first. This should discourage API users from using keywords from other data sources present only for backwards compatibility.

### CODE REVIEW

File: events/api.py [hunk 1]
```diff
@@ -562,36 +562,59 @@ class KeywordRetrieveViewSet(mixins.RetrieveModelMixin, viewsets.GenericViewSet)
...
+        queryset = queryset.filter(data_source=data_source)
+
         if not self.request.query_params.get('show_all_keywords'):
-            events = Event.objects.all()
+            queryset = queryset.filter(n_events__gt=0)
+            # if extra parameters are provided, they are applied on events:
             params = _clean_qp(self.request.query_params)
-            if 'data_source' in params:
-                del params['data_source']
-            events = _filter_event_queryset(events, params)
```
Comment (juyrjola) [inline events/api.py]: Remove this mystery

File: events/models.py [hunk 2]
```diff
@@ -204,24 +209,52 @@ class KeywordLabel(models.Model):
...
 class Keyword(BaseModel):
     alt_labels = models.ManyToManyField(KeywordLabel, blank=True, related_name='keywords')
     aggregate = models.BooleanField(default=False)
+    deprecated = models.BooleanField(default=False, db_index=True)
     objects = models.Manager()
+    n_events = models.IntegerField(
+        verbose_name=_('event count'),
+        help_text=_('number of events with this keyword'),
+        default=0,
+        editable=False
```
Comment (juyrjola) [inline events/models.py]: Add `db_index=True`

File: events/api.py [hunk 3]
```diff
@@ -835,7 +838,7 @@ class EventSerializer(LinkedEventsSerializer, GeoModelAPIView):
     # provider = OrganizationSerializer(hide_ld_context=True)
     keywords = JSONLDRelatedField(serializer=KeywordSerializer, many=True, allow_empty=False,
                                   required=False,
-                                  view_name='keyword-detail', queryset=Keyword.objects.all())
+                                  view_name='keyword-detail', queryset=Keyword.objects.filter(deprecated=False))
```
Comment (juyrjola) [inline events/api.py]: https://docs.djangoproject.com/en/1.10/topics/db/managers/#modifying-a-manager-s-initial-queryset ?

File: events/api.py [hunk 4]
```diff
@@ -563,36 +563,39 @@ class KeywordRetrieveViewSet(mixins.RetrieveModelMixin, viewsets.GenericViewSet)
...
+            queryset = queryset.filter(n_events__gt=0)
+        if not self.request.query_params.get('show_deprecated'):
+            queryset = queryset.filter(deprecated=False)
 
         # Optionally filter keywords by filter parameter,
         # can be used e.g. with typeahead.js
         val = self.request.query_params.get('filter')
         if val:
-            queryset = queryset.filter(name__startswith=val)
+            queryset = queryset.filter(name__contains=val)
```
Comment (juyrjola) [inline events/api.py]: Add a filter `text` instead with filter `icontains`

File: events/importer/sync.py [hunk 5]
```diff
@@ -37,6 +39,9 @@ def finish(self):
         if len(delete_list) > 5 and len(delete_list) > len(self.obj_dict) * 0.2:
             raise Exception("Attempting to delete more than 20% of total items")
         for obj in delete_list:
+            if self.allow_deleting_func:
+                if not self.allow_deleting_func(obj):
+                    raise Exception("Deleting " + str(obj) + " not allowed by the importer")
```
Comment (juyrjola) [inline events/importer/sync.py]: Use either the `%` operator or `str.format()`

File: events/importer/yso.py [hunk 6]
```diff
@@ -19,10 +19,31 @@
...
+}
+
 def is_deprecated(graph, subject):
     return (subject, OWL.deprecated, None) in graph
 def is_aggregate_concept(graph, subject):
     return (subject, SKOS.inScheme, rdflib.term.URIRef(yso+'aggregateconceptscheme')) in graph
+def allow_deprecating_keyword(keyword):
+    if keyword.events.all().exists() or keyword.audience_events.all().exists():
+        if keyword.id not in YSO_DEPRECATED_MAPS:
+            raise Exception("Deprecating YSO keyword " + str(keyword) + " that is referenced in events " +
```
Comment (juyrjola) [inline events/importer/yso.py]: `str.format()`

File: events/importer/yso.py [hunk 7]
```diff
@@ -102,17 +123,29 @@ def save_keywords(self, graph):
...
 
         if not bulk_mode:
-            queryset = Keyword.objects.filter(data_source=self.data_source)
+            deprecate_keyword = lambda obj: obj.deprecate()
+            check_deprecated_keyword = lambda obj: obj.deprecated
+            # manually add new keywords to deprecated ones
+            for old_id, new_id in YSO_DEPRECATED_MAPS.items():
+                old_keyword = Keyword.objects.get(id=old_id)
+                new_keyword = Keyword.objects.get(id=new_id)
+                print('Mapping events with ' + str(old_keyword) + ' to ' + str(new_keyword))
```
Comment (juyrjola) [inline events/importer/yso.py]: str.format()

File: events/models.py [hunk 8]
```diff
@@ -204,24 +209,52 @@ class KeywordLabel(models.Model):
...
         verbose_name_plural = _('keywords')
 
+
+def recache_n_events(keyword):
+    """
+    The helper function has to exist outside the model, because it is used in migration.
+    Django apps cannot serialize unbound instance functions when saving model history during migration.
+    :type keyword: Keyword
+    """
+    n_events = (keyword.events.all() | keyword.audience_events.all()).count()
```
Comment (juyrjola) [inline events/models.py]: `.distinct()`

**Comment by Rikuoja** [comment]
Fixed version of this merged in #187.

**Pull request closed without merging**